### ISSUE TITLE
Bitwise operators confuse null analysis

### ISSUE THREAD

**Issue by Yen-David** [dotnet/roslyn#40304]
**Version Used**: C# 8, VS 16.3.10

**Steps to Reproduce**:

1. Compile the code below:
```
  static bool BoxesNil(this object x) => object.ReferenceEquals(x, DBNull.Value);
  static bool BoxesNull(this object x) => x == null | x.BoxesNil();
```
2. You get a null warning for "x.BoxesNil()"!

**Expected Behavior**: The parameter x is already not nullable. Why trigger a null warning?
**Actual Behavior**: It does trigger a warning.

Flow analysis to safe-guard against nulls doesn't work too well with bitwise operators. The reason why null checks are still performed for non-nullable variables is to support interoperability with older versions of the .Net family, Standard or not. Likewise, bitwise operators are preferred instead of logical operators mostly because a jump/branch is too expensive, especially on older processors, nearing the cost of a callvirt. On newer processors, a jump/branch still nears the cost of a delegate call before JITting. Well, calls are mere jumps/branches. What do you expect anyway?

**Comment by jcouv** [comment]
Relates to https://github.com/dotnet/roslyn/issues/37344 (possible duplicate)

**Comment by jcouv** [comment]
Looking at this closer, the problem is not with the `|` operator, but with the declaration of your methods.
`x.BoxesNil()` won't produce a null reference exception, because it is an extension method which can handle `null`. You must declare that.

``` csharp
using System;
#nullable enable

public static class Extension
{
    public static bool BoxesNil(this object? x) => object.ReferenceEquals(x, DBNull.Value);
    public static bool BoxesNull(this object? x) => x == null | x.BoxesNil();
}

public class C
{
    public static void Main()
    {
        object? x = null;
        x.BoxesNull();
    }
}
```

[sharplab](https://sharplab.io/#v2:D4AQTAjAsAUAxAOwK4BsUEMBGKCmACHBLXWWEAZjxAgDYqw8BRADwBdCBnASwHsFYA3rDwiqlanUw8eKPACEezHBwByXFAApWACy4c8PTACscAY1YB+PMwCUeALwA+A8bOsAdACUcAMxwAnQlMcRgBHJHQUDg1mABo8ABE5FVQUdwA1SKQcGwBuYVEKKlo8KRl5RWUUtC1dfUMTcytbB2dmB3s8ZDQ8YGt3BSVVdQ082ABfUhgi8DwAYUECkXAwJbwhGFEtl0bLawcu1PzN7ZFmAcrVVNHjrcmYcaA==)

I think the compiler is behavior by-design, so I'll close the issue. Please re-open if you believe otherwise.

**Comment by Yen-David** [comment]
Well, there's definitely a logic error in the null analysis performed by the compiler. I don't have the time to look into further. However, if you are talking about workarounds, there are plenty. Declare the single checks as follows:

```
static bool BoxesDBNull(this object x) => object.ReferenceEquals(x, DBNull.Value);
static bool BoxesNull(this object? x) => x == null;
```
Now, you can define the double check BoxesNil() many ways to work around this issue:

```
static bool BoxesNil(this object x) => x.BoxesDBNull() | x == null;
static bool BoxesNil(this object x) => x.BoxesNull() | x.BoxesDBNull();
```

Basically, you can call x.BoxesNull with bitwise operators and x == null with logical operators for full generality in the current behavior by design.

**Comment by Yen-David** [comment]
@jcouv, just to make sure that you get notified. I don't know if updating the comment does. =)

**Comment by jcouv** [comment]
> Well, there's definitely a logic error in the null analysis performed by the compiler.

I fail to see it.  The behavior is what I expect. In `x == null | x.ToString()` the compiler would warn of a possible dereference (`x.ToString()` would cause an NRE), and similarly in the case of an extension method that only accepts a non-nullable `object` the compiler should warn about passing a maybe-null value.
What I suggested (declare methods as accepting `object?` instead of `object`) is not a workaround, but the proper code as far as I can tell (tells the compiler that those methods accept null inputs).

I hope that makes sense.

**Comment by Yen-David** [comment]
@jcouv, bitwise operators are rarely useful, mostly as workarounds where compilers fail to optimize. In worse cases, falling back to C/C++ or even assembly is the way to go. I say the bug is not very important. Logical operators and all jumps/branches alike can be treated as inline candidates, such that bitwise operators will be restricted to integers.

In your example, ToString() and anything that allocates memory are too expensive, overwhelming the cost of jumps/branches. Plus, ?. and ?[] are very fast and efficient, no need for bitwise operators. Jumps/branches work against processor cache and instruction reordering/parallelization, the primary reason why they are expensive but not as bad as we would think.

As I already mentioned, "object" is declared to guard against nulls at compile-time, but null checks are still performed to guard against nulls at run-time for interoperability. That's all. My hope is to create some sort of a visual language where we no longer have to deal with syntactical back-compatibility, 2D or 3D graph models (vertical stacks and horizontal queues as two different dimensions) instead of 1D long strings. C# is getting too big by now, close enough to C++ with a GC.

I don't mind closing this bug, because I know my voice has been heard.
